### PR TITLE
1873: tile laying rules

### PR DESCRIPTION
**Pull request by roseundy**
Common file changes:
* UI::Round::Stock - allow minors as parameter to Action::Merge
* Action::DestinationConnection - allow minors as parameters
* Step::Tracker - call game method for graph to check and clear
* Game::Base - use init_graph method instead of calling Graph.new directly; implement method for above

1873 implementation:
* fix public mine formation via SR view
* independent mines can lay one tile if not connected
* check for independent mine connection after every tile lay (mine or RR)
* concession route tile lays for concession company
* reserving tiles when other company lays on concession route
* reimburse owner when laying on other company's concession route (and back charge concession company when they complete the concession route)
* limit concession route tile lays to required tiles only (possibly from reserved list)
* concession route tile lay of reserved tile can be any color
* free tile reservations when needed
* prohibit laying on "locked" hexes until phase 3
* RR can lay 2 tiles (after concession route is done) if both are either yellow or green with 2 unconnected sections of track
* towns upgrade to cities
* bordered tiles can only upgrade to bordered tiles

### CODE REVIEW

File: assets/app/view/game/round/stock.rb [hunk 1]
```diff
@@ -83,10 +83,7 @@ def render_merge_button
           merge = lambda do
             if @selected_corporation
               do_merge = lambda do
-                process_action(Engine::Action::Merge.new(
-                  @mergeable_entity,
-                  corporation: @selected_corporation,
-                ))
+                merge_action(@mergeable_entity, @selected_corporation)
```
Comment (tobymao) [inline assets/app/view/game/round/stock.rb:86]: z = corporation? ? { corporation: x } : { minor: x}
process_action(Action.new(merge_entity, **z)

File: lib/engine/game/g_1873/game.rb [hunk 2]
```diff
@@ -97,6 +152,21 @@ def setup
...
+          @reserved_tiles = Hash.new { |h, k| h[k] = {} }
+          @state_network_hexes = STATE_NETWORK.map { |h| hex_by_id(h) }
+        end
+
+        def init_graph
+          Graph.new(self, skip_track: :broad)
+        end
+
+        def graph_for_entity(entity)
+          if entity.minor?
```
Comment (tobymao) [inline lib/engine/game/g_1873/game.rb:165]: one line with a ternary

File: lib/engine/game/g_1873/step/form.rb [hunk 3]
```diff
@@ -10,7 +10,7 @@ class Form < Engine::Step::Base
           MERGE_ACTION = %w[merge].freeze
 
           def actions(entity)
-            return [] unless entity == owner
+            return [] unless entity == owner || entity == buyer
```
Comment (tobymao) [inline lib/engine/game/g_1873/step/form.rb:13]: change to an if statement

File: assets/app/view/game/round/stock.rb (hunk 1 above)
Comment (roseundy) [reply assets/app/view/game/round/stock.rb:86]: OK.

File: lib/engine/game/g_1873/game.rb (hunk 2 above)
Comment (roseundy) [reply lib/engine/game/g_1873/game.rb:165]: OK

File: lib/engine/game/g_1873/step/form.rb (hunk 3 above)
Comment (roseundy) [reply lib/engine/game/g_1873/step/form.rb:13]: OK

**Pull request merged**